### PR TITLE
HPCC-29479 Support flexible secure user name comparison rules

### PR DESCRIPTION
**Pull request by timothyklemm**
Enable secure user name matching based on security-manager defined rules. Define the default rule to be a case-insensitive comparison, while allowing individual security managers to customize behavior by manipulating the user objects they construct.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/security/shared/SecureUser.hpp [hunk 1]
```diff
@@ -328,6 +328,11 @@ class CSecureUser : implements ISecUser, implements ISecCredentials, public CInt
     {
         return false;
     }
+
+    virtual bool isCanonicalMatch(const char* name) const override
+    {
+        return (name && strieq(m_name, name));
```
Comment (RussWhitehead) [inline system/security/shared/SecureUser.hpp:334]: m_name is never tested to ensure it isn't null. Not sure if that's an exposure here, but consider adding a check in the ctor above, or at least a check here for n_name not null or empty

File: system/security/shared/SecureUser.hpp [hunk 2]
```diff
@@ -328,6 +328,11 @@ class CSecureUser : implements ISecUser, implements ISecCredentials, public CInt
     {
         return false;
     }
+
+    virtual bool isCanonicalMatch(const char* name) const override
```
Comment (RussWhitehead) [inline system/security/shared/SecureUser.hpp:332]: I think a comment on what this function does is warranted. Future SecMgr develops will thank you !

Comment (timothyklemm) [reply system/security/shared/SecureUser.hpp:332]: Will add the first comment to the interface, so future developers don't have to track down all implementations to figure it out.

File: system/security/shared/SecureUser.hpp (hunk 1 above)
Comment (afishbeck) [reply system/security/shared/SecureUser.hpp:334]: Either way is fine, but m_name.str() will be easier to read.

**Review by afishbeck — COMMENTED** [review]
Looks fine except for the null issue and the requested code comment.

**Review by RussWhitehead — APPROVED** [review]
@timothyklemm Looks good

**Review by afishbeck — APPROVED** [review]

**Comment by timothyklemm** [comment]
@ghalliday please merge.

**Pull request merged**